[folly,proxygen,snappy,fbthrift,cachelib] Fix integration

dg0yt opened the pull request:
Alternative to #44669.
Revise dependency control/import/export.
Replace __has_include where vcpkg requires precise control.
Add rtti to snappy, needed by folly.

dg0yt committed:
[folly] Revise and fix deps, cleanup

Leverage VCPKG_LOCK_FIND_PACKAGE.

dg0yt committed:
Fixup

dg0yt committed:
Fixup

JonLiu1993 commented:
@dg0yt, when I tested the PR on my machine, I still got the ci error：
````
CMake Error at scripts/cmake/vcpkg_execute_build_process.cmake:134 (message):
    Command failed: /home/vcpkg/folly/downloads/tools/cmake-3.30.1-linux/cmake-3.30.1-linux-x86_64/bin/cmake --build . --config Debug --target install -- -v -j3
    Working Directory: /home/vcpkg/folly/buildtrees/proxygen/x64-linux-dbg
    See logs for more information:
      /home/vcpkg/folly/buildtrees/proxygen/install-x64-linux-dbg-out.log

Call Stack (most recent call first):
  installed/x64-linux/share/vcpkg-cmake/vcpkg_cmake_build.cmake:74 (vcpkg_execute_build_process)
  installed/x64-linux/share/vcpkg-cmake/vcpkg_cmake_install.cmake:16 (vcpkg_cmake_build)
  ports/proxygen/portfile.cmake:23 (vcpkg_cmake_install)
  scripts/ports.cmake:203 (include)
````
/home/vcpkg/folly/buildtrees/proxygen/install-x64-linux-dbg-out.log
````
/home/vcpkg/folly/buildtrees/proxygen/src/5.03.24.00-3be2db0ed2.clean/proxygen/lib/services/WorkerThread.cpp:33:14: error: ‘folly::PollIoBackend’ has not been declared
   33 |       folly::PollIoBackend::Options options;
      |              ^~~~~~~~~~~~~
/home/vcpkg/folly/buildtrees/proxygen/src/5.03.24.00-3be2db0ed2.clean/proxygen/lib/services/WorkerThread.cpp:34:7: error: ‘options’ was not declared in this scope; did you mean ‘optind’?
   34 |       options.setCapacity(static_cast<size_t>(FLAGS_pwt_io_uring_capacity))
      |       ^~~~~~~
      |       optind

````
Steps to reproduce：
1. ./vcpkg install liburing: x64-linux
2. ./vcpkg install proxygen:x64-linux

dg0yt committed:
Adjust gflags and zlib

gflags: Put target namespace requirement into export.
zlib: Use target, as in proxygen.

dg0yt committed:
[proxygen] Revise dependencies

dg0yt committed:
Merge remote-tracking branch 'origin/master' into folly

dg0yt committed:
versions

BillyONeal commented:
If this lands https://github.com/microsoft/vcpkg/pull/44753 should be closed. If that one lands first, undoing it goes here.

dg0yt commented:
It is ready for review.

LilyWangLL commented:
There are new errors of installing `proxygen:x64-linux`:
```
[175/186] : && /usr/bin/c++ -fPIC -g  proxygen/httpserver/CMakeFiles/proxygen_push.dir/samples/push/PushServer.cpp.o proxygen/httpserver/CMakeFiles/proxygen_push.dir/samples/push/PushRequestHandler.cpp.o -o proxygen/httpserver/proxygen_push  proxygen/lib/libproxygen.a  proxygen/httpserver/libproxygenhttpserver.a  proxygen/lib/libproxygen.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libwangle.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libboost_iostreams.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libbz2d.a  /mnt/lily/pro/installed/x64-linux/debug/lib/liblzma.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libboost_random.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_fizz_client.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_client.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_server.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_server_state.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_transport.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_batch_writer.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_cc_algo.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_codec_pktrebuilder.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_ack_scheduler.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_happyeyeballs.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_async_udp_socket.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_events.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_looper.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_observer.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_state_datagram_handler.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_state_pacing_functions.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_buf_accessor.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_state_ack_handler.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_loss.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_state_simple_frame_functions.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_state_functions.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_codec_pktbuilder.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_state_stream.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_flowcontrol.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_qlogger.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_state_stream_functions.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_fizz_handshake.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_dsr_frontend.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_dsr_types.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_transport_knobs.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_transport_settings_functions.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_state_machine.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_codec.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_handshake.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_codec_packet_number_cipher.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_codec_decode.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_codec_types.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_bufutil.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_constants.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_exception.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libfizz.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libfolly.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libboost_filesystem.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libboost_thread.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libboost_atomic.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libboost_chrono.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libboost_date_time.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libssl.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libcrypto.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libevent_extrad.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libboost_context.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libboost_program_options.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libboost_container.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libboost_system.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libfmtd.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libdouble-conversion.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libevent_pthreadsd.a  /mnt/lily/pro/installed/x64-linux/lib/libbz2.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libbz2d.a  /mnt/lily/pro/installed/x64-linux/debug/lib/liblz4d.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libsnappy.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libaio.a  /mnt/lily/pro/installed/x64-linux/debug/lib/liburing.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libunwind.a  /mnt/lily/pro/installed/x64-linux/debug/lib/liblzma.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libz.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libssl.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libcrypto.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libsodium.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libglog.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libgflags_debug.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libzstd.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libevent_cored.a  -ldl  /usr/lib/x86_64-linux-gnu/librt.a && :
FAILED: proxygen/httpserver/proxygen_push
: && /usr/bin/c++ -fPIC -g  proxygen/httpserver/CMakeFiles/proxygen_push.dir/samples/push/PushServer.cpp.o proxygen/httpserver/CMakeFiles/proxygen_push.dir/samples/push/PushRequestHandler.cpp.o -o proxygen/httpserver/proxygen_push  proxygen/lib/libproxygen.a  proxygen/httpserver/libproxygenhttpserver.a  proxygen/lib/libproxygen.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libwangle.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libboost_iostreams.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libbz2d.a  /mnt/lily/pro/installed/x64-linux/debug/lib/liblzma.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libboost_random.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_fizz_client.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_client.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_server.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_server_state.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_transport.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_batch_writer.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_cc_algo.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_codec_pktrebuilder.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_ack_scheduler.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_happyeyeballs.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_async_udp_socket.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_events.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_looper.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_observer.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_state_datagram_handler.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_state_pacing_functions.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_buf_accessor.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_state_ack_handler.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_loss.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_state_simple_frame_functions.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_state_functions.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_codec_pktbuilder.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_state_stream.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_flowcontrol.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_qlogger.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_state_stream_functions.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_fizz_handshake.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_dsr_frontend.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_dsr_types.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_transport_knobs.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_transport_settings_functions.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_state_machine.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_codec.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_handshake.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_codec_packet_number_cipher.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_codec_decode.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_codec_types.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_bufutil.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_constants.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libmvfst_exception.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libfizz.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libfolly.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libboost_filesystem.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libboost_thread.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libboost_atomic.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libboost_chrono.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libboost_date_time.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libssl.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libcrypto.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libevent_extrad.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libboost_context.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libboost_program_options.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libboost_container.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libboost_system.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libfmtd.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libdouble-conversion.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libevent_pthreadsd.a  /mnt/lily/pro/installed/x64-linux/lib/libbz2.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libbz2d.a  /mnt/lily/pro/installed/x64-linux/debug/lib/liblz4d.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libsnappy.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libaio.a  /mnt/lily/pro/installed/x64-linux/debug/lib/liburing.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libunwind.a  /mnt/lily/pro/installed/x64-linux/debug/lib/liblzma.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libz.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libssl.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libcrypto.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libsodium.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libglog.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libgflags_debug.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libzstd.a  /mnt/lily/pro/installed/x64-linux/debug/lib/libevent_cored.a  -ldl  /usr/lib/x86_64-linux-gnu/librt.a && :
/usr/bin/ld: /mnt/lily/pro/installed/x64-linux/debug/lib/libfolly.a(Compression.cpp.o):(.data.rel.ro+0x3c0): undefined reference to `typeinfo for snappy::Source'
collect2: error: ld returned 1 exit status
```

The steps to reproduce this error as below:

1. `./vcpkg install folly[snappy]:x64-linux`
2. `./vcpkg install proxygen:x64-linux`

dg0yt commented:
Hm, libs are complete and in traditional order:
`... /mnt/lily/pro/installed/x64-linux/debug/lib/libfolly.a ... /mnt/lily/pro/installed/x64-linux/debug/lib/libsnappy.a ...`
Probably not even a regression?

dg0yt commented:
https://github.com/facebook/folly/issues/1606

dg0yt commented:
https://github.com/facebook/proxygen/issues/361

dg0yt committed:
[vcpkg-ci-folly] New test port

dg0yt commented:
And also `thrift` has a problem with `folly[liburing]`:
~~~
FAILED: thrift/lib/cpp2/CMakeFiles/thriftcpp2.dir/security/SSLUtil.cpp.o 
/usr/bin/c++ -DBOOST_ATOMIC_NO_LIB -DBOOST_ATOMIC_STATIC_LINK -DBOOST_CHRONO_NO_LIB -DBOOST_CHRONO_STATIC_LINK -DBOOST_CONTAINER_NO_LIB -DBOOST_CONTAINER_STATIC_LINK -DBOOST_CONTEXT_EXPORT="" -DBOOST_CONTEXT_NO_LIB="" -DBOOST_CONTEXT_STATIC_LINK="" -DBOOST_DATE_TIME_NO_LIB -DBOOST_DATE_TIME_STATIC_LINK -DBOOST_FILESYSTEM_NO_LIB -DBOOST_FILESYSTEM_STATIC_LINK=1 -DBOOST_PROGRAM_OPTIONS_NO_LIB -DBOOST_PROGRAM_OPTIONS_STATIC_LINK -DBOOST_THREAD_NO_LIB -DBOOST_THREAD_STATIC_LINK -DBOOST_THREAD_USE_LIB -DGFLAGS_IS_A_DLL=0 -DGLOG_USE_GFLAGS -DGLOG_USE_GLOG_EXPORT -DTHRIFT_HAVE_LIBSNAPPY=0 -I/mnt/vcpkg-ci/b/fbthrift/src/5.03.31.00-df4db1c197.clean/. -I/mnt/vcpkg-ci/b/fbthrift/x64-linux-dbg -isystem /mnt/vcpkg-ci/installed/x64-linux/include -fPIC -g -std=c++17 -fcoroutines -fsized-deallocation -MD -MT thrift/lib/cpp2/CMakeFiles/thriftcpp2.dir/security/SSLUtil.cpp.o -MF thrift/lib/cpp2/CMakeFiles/thriftcpp2.dir/security/SSLUtil.cpp.o.d -o thrift/lib/cpp2/CMakeFiles/thriftcpp2.dir/security/SSLUtil.cpp.o -c /mnt/vcpkg-ci/b/fbthrift/src/5.03.31.00-df4db1c197.clean/thrift/lib/cpp2/security/SSLUtil.cpp
/mnt/vcpkg-ci/b/fbthrift/src/5.03.31.00-df4db1c197.clean/thrift/lib/cpp2/security/SSLUtil.cpp: In function ‘folly::AsyncSocketTransport::UniquePtr apache::thrift::moveToPlaintext(FizzSocket*)’:
/mnt/vcpkg-ci/b/fbthrift/src/5.03.31.00-df4db1c197.clean/thrift/lib/cpp2/security/SSLUtil.cpp:138:56: error: ‘AsyncIoUringSocket’ is not a member of ‘folly’
  138 |       fizzSock->template getUnderlyingTransport<folly::AsyncIoUringSocket>()) {
      |                                                        ^~~~~~~~~~~~~~~~~~
/mnt/vcpkg-ci/b/fbthrift/src/5.03.31.00-df4db1c197.clean/thrift/lib/cpp2/security/SSLUtil.cpp:147:12: error: ‘folly::AsyncIoUringSocket’ has not been declared
  147 |     folly::AsyncIoUringSocket::UniquePtr io =
      |            ^~~~~~~~~~~~~~~~~~
/mnt/vcpkg-ci/b/fbthrift/src/5.03.31.00-df4db1c197.clean/thrift/lib/cpp2/security/SSLUtil.cpp:150:9: error: ‘io’ was not declared in this scope; did you mean ‘folly::io’?
  150 |     if (io) {
      |         ^~
      |         folly::io
In file included from /mnt/vcpkg-ci/installed/x64-linux/include/folly/io/Cursor.h:33,
                 from /mnt/vcpkg-ci/installed/x64-linux/include/fizz/record/Types.h:12,
                 from /mnt/vcpkg-ci/installed/x64-linux/include/fizz/client/ClientExtensions.h:11,
                 from /mnt/vcpkg-ci/installed/x64-linux/include/fizz/client/AsyncFizzClient.h:11,
                 from /mnt/vcpkg-ci/b/fbthrift/src/5.03.31.00-df4db1c197.clean/./thrift/lib/cpp2/security/SSLUtil.h:19,
                 from /mnt/vcpkg-ci/b/fbthrift/src/5.03.31.00-df4db1c197.clean/thrift/lib/cpp2/security/SSLUtil.cpp:17:
/mnt/vcpkg-ci/installed/x64-linux/include/folly/io/IOBufQueue.h:27:11: note: ‘folly::io’ declared here
   27 | namespace io {
      |           ^~
~~~

dg0yt committed:
[fbthrift] Wire FOLLY_HAS_LIBURING

dg0yt committed:
[snappy] Add rtti feature

dg0yt committed:
[folly] Use snappy[rtti]

dg0yt committed:
Merge remote-tracking branch 'origin/master' into folly

dg0yt committed:
[vcpkg baseline] Restore proxygen:x64-linux

dg0yt committed:
[cachelib] Disable uring

dg0yt committed:
v

dg0yt commented on ports/snappy/portfile.cmake:
```diff
@@ -7,12 +7,19 @@ vcpkg_from_github(
     PATCHES
         no-werror.patch
         pkgconfig.diff
+        rtti.diff
```
Similar in condaforge.  
Won't upstream, because it was already rejected for lack-of-interest.  
https://github.com/google/snappy/pull/147

vicroms commented:
Thanks for the PR!
